core/src/find_packet_sender_stake_stage.rs

```diff
@@ -0,0 +1,155 @@
+            );
+            *self = FindPacketSenderStakeStats::default();
+            self.last_print = now;
+        }
+    }
+}
+
+pub struct FindPacketSenderStakeStage {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl FindPacketSenderStakeStage {
+    pub fn new(
+        packet_receiver: FindPacketSenderStakeReceiver,
+        sender: FindPacketSenderStakeSender,
+        bank_forks: Arc<RwLock<BankForks>>,
+        cluster_info: Arc<ClusterInfo>,
+    ) -> Self {
+        let mut stats = FindPacketSenderStakeStats::default();
+        let cluster_nodes_cache = Arc::new(ClusterNodesCache::<FindPacketSenderStakeStage>::new(
+            CLUSTER_NODES_CACHE_NUM_EPOCH_CAP,
+            CLUSTER_NODES_CACHE_TTL,
+        ));
+        let thread_hdl = Builder::new()
+            .name("find-packet-sender-stake".to_string())
+            .spawn(move || {
+                let mut last_stakes = Instant::now();
+                let mut ip_to_stake: HashMap<IpAddr, u64> = HashMap::new();
+                loop {
+                    let mut refresh_ip_to_stake_time = Measure::start("refresh_ip_to_stake_time");
+                    if last_stakes.elapsed().as_millis() > STAKES_REFRESH_PERIOD_IN_MS {
+                        let (root_bank, working_bank) = {
+                            let bank_forks = bank_forks.read().unwrap();
+                            (bank_forks.root_bank(), bank_forks.working_bank())
+                        };
+                        ip_to_stake = cluster_nodes_cache
+                            .get(root_bank.slot(), &root_bank, &working_bank, &cluster_info)
+                            .get_ip_to_stakes();
+                        last_stakes = Instant::now();
+                    }
+                    refresh_ip_to_stake_time.stop();
+                    stats.refresh_ip_to_stake_time = stats
+                        .refresh_ip_to_stake_time
+                        .saturating_add(refresh_ip_to_stake_time.as_us());
+
+                    match streamer::recv_packet_batches(&packet_receiver) {
+                        Ok((mut batches, _num_packets, _recv_duration)) => {
+                            let mut apply_sender_stakes_time =
+                                Measure::start("apply_sender_stakes_time");
+                            Self::apply_sender_stakes(&mut batches, &ip_to_stake);
+                            apply_sender_stakes_time.stop();
+                            stats.apply_sender_stakes_time = stats
+                                .apply_sender_stakes_time
+                                .saturating_add(apply_sender_stakes_time.as_us());
+                            let mut send_batches_time = Measure::start("send_batches_time");
+                            if let Err(e) = sender.send(batches) {
+                                info!("Sender error: {:?}", e);
+                            }
+                            send_batches_time.stop();
```

Comment:
nit: One thing we've started doing is using `Measure::this()` https://github.com/solana-labs/solana/blob/46dfbc91ce6f004a46e8d9695bd0bca23eea846e/core/src/banking_stage.rs#L589-L593 to make some of these measurement code blocks clearer